pkg/apis/provisioning/v1alpha2/provisioner_defaults.go

```diff
@@ -16,10 +16,67 @@ package v1alpha2
 
 import (
 	"context"
+	"encoding/base64"
+	"errors"
+	"io/ioutil"
+	"os"
+
+	"go.uber.org/zap"
 )
 
 // SetDefaults for the provisioner
 func (p *Provisioner) SetDefaults(ctx context.Context) {}
 
 // SetDefaults for the provisioner, cascading to all subspecs
 func (s *ProvisionerSpec) SetDefaults(ctx context.Context) {}
+
+// WithDefaults returns a copy of this Provisioner with some empty/missing
+// properties replaced by (potentially dynamic) cloud provider agnostic default values.
+// The returned copy might be complemented by dynamic default values which
+// must not be hoisted (saved) into the original Provisioner CRD as those
+// default values might change over time (e.g. rolling upgrade of CABundle, ...).
+func (p *Provisioner) WithDynamicDefaults() (_ Provisioner, err error) {
+	provisioner := *p.DeepCopy()
+	provisioner.Spec, err = provisioner.Spec.withDynamicDefaults()
+	return provisioner, err
+}
+
+// WithDefaults returns a copy of this ProvisionerSpec with some empty
+// properties replaced by default values.
+func (s *ProvisionerSpec) withDynamicDefaults() (_ ProvisionerSpec, err error) {
+	spec := *s.DeepCopy()
+	spec.Cluster, err = spec.Cluster.withDynamicDefaults()
+	return spec, err
+}
+
+// WithDefaults returns a copy of this Cluster with some empty
+// properties replaced by default values. Notably, it will try
+// to load the CABundle from the in-cluster configuraiton if it
+// is not explicitly set.
+func (c *Cluster) withDynamicDefaults() (_ Cluster, err error) {
+	cluster := *c.DeepCopy()
+	cluster.CABundle, err = cluster.getCABundle()
+	return cluster, err
+}
+
+func (c *Cluster) getCABundle() (*string, error) {
+	if c.CABundle != nil {
+		// If CABundle is explicitly provided use that one. An empty string is
+		// a valid value here if no CABundle is required at all.
+		return c.CABundle, nil
+	} else {
+		// Otherwise, fallback to the in-cluster configuration.
+		if binary, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"); err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				zap.S().Infof("in-cluster CABundle file /var/run/secrets/kubernetes.io/serviceaccount/ca.crt not found, %w", err)
+				//TODO shall this be treated as an error or is it a valid use case that a cluster has no CABundle at all?
+				return nil, nil
+			}
+			return nil, err
+		} else {
+			zap.S().Debugf("using in-cluster CABundle from file /var/run/secrets/kubernetes.io/serviceaccount/ca.crt, %s", string(binary))
+			encoded := base64.StdEncoding.EncodeToString(binary)
+			return &encoded, nil
```

Comment:
Any reason we're returning this as a string pointer? Seems like it creates NPE opportunities that could be more safely handled by using an empty string for the CABundle if something goes wrong. It also allows you to simplify the code and use `return base64.StdEncoding.EncodeToString(binary), nil`

Reply:
This is because (as you suggested) I refactored `Clutser.CABundle` to type `*string`. Of course I can change the return type fo this function to `string` but that will just move the need for an intermediate variable (to get a pointer) to the calling function.

Reply:
Oopsie you're 100% correct.